drivers/vulkan/SCsub

```diff
@@ -2,124 +2,72 @@
 
 Import("env")
 
+if env["use_volk"]:
+    env.AppendUnique(CPPDEFINES=["USE_VOLK"])
+
 thirdparty_obj = []
+thirdparty_volk_dir = "#thirdparty/volk"
 
-# FIXME: Refactor all this to reduce code duplication.
 if env["platform"] == "android":
-    # Use NDK Vulkan headers
-    thirdparty_dir = env["ANDROID_NDK_ROOT"] + "/sources/third_party/vulkan/src"
-    thirdparty_includes = [
-        thirdparty_dir,
-        thirdparty_dir + "/include",
-        thirdparty_dir + "/layers",
-        thirdparty_dir + "/layers/generated",
-    ]
-    env.Prepend(CPPPATH=thirdparty_includes)
-
-    # Build Vulkan memory allocator
+    if env["use_volk"]:
+        # Use bundled Vulkan headers
+        thirdparty_dir = "#thirdparty/vulkan"
+        env.Prepend(CPPPATH=[thirdparty_volk_dir, thirdparty_dir, thirdparty_dir + "/include"])
+    else:
+        # Use NDK Vulkan headers
+        thirdparty_dir = env["ANDROID_NDK_ROOT"] + "/sources/third_party/vulkan/src"
+        thirdparty_includes = [
+            thirdparty_dir,
+            thirdparty_dir + "/include",
+            thirdparty_dir + "/layers",
+            thirdparty_dir + "/layers/generated",
+        ]
+        env.Prepend(CPPPATH=thirdparty_includes)
+
+    env.AppendUnique(CPPDEFINES=["VK_USE_PLATFORM_ANDROID_KHR"])
+
+    # Build Vulkan memory allocator and volk
     env_thirdparty = env.Clone()
     env_thirdparty.disable_warnings()
 
     thirdparty_dir = "#thirdparty/vulkan"
-    vma_sources = [thirdparty_dir + "/android/vk_mem_alloc.cpp"]
-    env_thirdparty.add_source_files(thirdparty_obj, vma_sources)
 
-elif env["platform"] == "iphone":
-    # Use bundled Vulkan headers
-    thirdparty_dir = "#thirdparty/vulkan"
-    env.Prepend(CPPPATH=[thirdparty_dir, thirdparty_dir + "/include", thirdparty_dir + "/loader"])
+    if env["use_volk"]:
+        env_thirdparty.AppendUnique(CPPDEFINES=["VMA_STATIC_VULKAN_FUNCTIONS=1"])
+        vma_sources = [thirdparty_dir + "/vk_mem_alloc.cpp", thirdparty_volk_dir + "/volk.c"]
```

Comment:
Based on https://github.com/zeux/volk#basic-usage:

> If some files in your application include `vulkan/vulkan.h` and don't include `volk.h`, this can result in symbol conflicts; consider defining `VK_NO_PROTOTYPES` when compiling code that uses Vulkan to make sure this doesn't happen.

and:

> https://github.com/zeux/volk/blob/master/volk.h#L13-L15

Doesn't that mean that it should work provided that we include `volk.h` **before** including `vulkan/vulkan.h`? Or if we define `VK_NO_PROTOTYPES`?

Reply:
But also, if it works, we might be able to get rid of the `android/vk_mem_alloc.cpp` now and just use the vendored one? It seems the custom one was just to get rid of some warnings, there might be a better way to do it? https://github.com/godotengine/godot/pull/37729#discussion_r406477853

Reply:
> Doesn't that mean that it should work provided that we include `volk.h` before including `vulkan/vulkan.h`? Or if we define `VK_NO_PROTOTYPES`?

BTW I think that means you don't need to patch thirdparty code to use volk, you just need to define `VK_NO_PROTOTYPES` in their build env I think.

Reply:
> This repository contains the Vulkan loader that is used for Linux, Windows, MacOS, and iOS. There is also a separate loader, maintained by Google, which is used on Android.

Seems like Android uses its custom loader, but we probably do not need to use NDK VMA. It's building fine with built-in one, and was added only to suppress some warnings, there are none now. I'll replace it with the built-on one and keep only custom loader headers. That's should fix #51524 build issues as well.

Reply:
> `VK_NO_PROTOTYPES` in their build env I think. Or we have to make sure to include `volk.h` before including them.

Neither of this works, it's failing with "error: redefinition of ... as different kind of symbol" errors.

Reply:
I guess that means that there's still a file which includes `vulkan/vulkan.h` before `volk.h` or before a definition of `VK_NO_PROTOTYPES`. But it's not a big deal, we can have another look at this later, or see with upstream if indeed what they described in the README no longer works with the current SDK version.